File: src/algorithm/sorting/bubble-sort/bubbleSort.js
#! /usr/local/bin/node


class BubbleSort {
    sort(originalList=[]) {
        let list = [...originalList];
        if (list.length <= 1) {
            return list;
        }

        for (let i = 0; i < list.length - 1; i++) {
            for (let j = 0; j < list.length - 1 - i; j++) {
                if (list[j] > list[j + 1]) {
                    [list[j], list[j + 1]] = [list[j + 1], list[j]];
                }
            }
        }

        return list;
    }

    isSorted(array=[]) {
        if (array.length <= 1) {
            return true;
        }

        for (let i = 1; i < array.length; i++) {
            if (array[i] < array[i - 1]) {
                return false;
            }
        }

        return true;
    }

    test() {
        const array = new Array(40000).fill(undefined).map(item => Math.random() * 10000);
        console.time('sort');
        let result = this.sort(array);
        console.timeEnd('sort');
        let sorted = this.isSorted(result);
        console.log('sorted:%s', sorted);
    }
}

new BubbleSort().test();

File: src/algorithm/sorting/bubble-sort/bubble_sort.py
#! /usr/local/bin/python3


class BubbleSort:
    def sort(self, original_list):
        arr = original_list.copy()

        if len(arr) <= 1:
            return arr

        for i in range(len(arr) - 1):
            for j in range(len(arr) - 1 - i):
                if arr[j] > arr[j + 1]:
                    arr[j], arr[j + 1] = arr[j + 1], arr[j]

        return arr

    def test(self):
        import time

        start = time.process_time()
        result = self.sort([3, 54, 54, 6, 32, 5, 8, 89, 23, 5, 64, 6, 7, 90, 3])
        end = time.process_time()

        print(result, end - start)

if __name__ == '__main__':
    BubbleSort().test()
